src/compiler/checker.ts

```diff
@@ -33694,8 +33694,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
 
         function getInstantiatedSignatures(signatures: readonly Signature[]) {
-            const applicableSignatures = filter(signatures, sig => !!sig.typeParameters && hasCorrectTypeArgumentArity(sig, typeArguments));
-            return sameMap(applicableSignatures, sig => {
+            const sameTypeAritySignatures = filter(signatures, sig => !!sig.typeParameters && hasCorrectTypeArgumentArity(sig, typeArguments));
+            if (!sameTypeAritySignatures.length) {
+                return sameTypeAritySignatures;
+            }
+            const applicableSignatures = mapDefined(sameTypeAritySignatures, sig => {
+                const typeArgumentTypes = checkTypeArguments(sig, typeArguments!, /*reportErrors*/ false);
+                return typeArgumentTypes && getSignatureInstantiation(sig, typeArgumentTypes, isInJSFile(sig.declaration));
+            });

```

Comment:
Note that this still doesn't exactly work like regular overload selection done by `chooseOverload`. I simply reject signatures that have non-matching constraints on their type params. When overloads are chosen for call expressions etc the first matching overload "wins". 

I think that this is actually desirable - since here we still output an intersection of signatures (overloads) and we leave the final overload selection for the time the output type is being used as a call expression.